Userland/Libraries/LibWeb/HTML/Worker.h

```diff
@@ -0,0 +1,260 @@
+
+    static DOM::ExceptionOr<NonnullRefPtr<Worker>> create(String& script_url, WorkerOptions const& options, DOM::Document& document);
+    static DOM::ExceptionOr<NonnullRefPtr<Worker>> create_with_global_object(Bindings::WindowObject& window, String& script_url, WorkerOptions const& options)
+    {
+        return Worker::create(script_url, options, window.impl().associated_document());
+    }
+
+    DOM::ExceptionOr<void> terminate();
+
+    void post_message(JS::Value& message, JS::Value& transfer);
+
+    virtual ~Worker() = default;
+
+    // ^EventTarget
+    virtual void ref_event_target() override { ref(); }
+    virtual void unref_event_target() override { unref(); }
+    virtual JS::Object* create_wrapper(JS::GlobalObject&) override;
+
+    MessagePort* implicit_message_port() { return m_implicit_port; }
+    RefPtr<MessagePort> outside_message_port() { return m_outside_port; }
+
+#undef __ENUMERATE
+#define __ENUMERATE(attribute_name, event_name)                  \
+    void set_##attribute_name(Optional<Bindings::CallbackType>); \
+    Bindings::CallbackType* attribute_name();
+    ENUMERATE_WORKER_EVENT_HANDLERS(__ENUMERATE)
+#undef __ENUMERATE
+
+protected:
+    //Worker(String&, WorkerOptions const&);
+    Worker(String&, WorkerOptions const&, DOM::Document&);
+
+private:
+    static HTML::EventLoop& get_vm_event_loop(JS::VM& target_vm)
+    {
+        return static_cast<Bindings::WebEngineCustomData*>(target_vm.custom_data())->event_loop;
+    }
+
+    String& m_script_url;
+    WorkerOptions m_options;
+    WeakPtr<DOM::Document> m_document;
+    Bindings::WebEngineCustomData m_custom_data;
+
+    NonnullRefPtr<JS::VM> m_worker_vm;
+    NonnullOwnPtr<JS::Interpreter> m_interpreter;
+    WeakPtr<WorkerEnvironmentSettingsObject> m_inner_settings;
+    JS::VM::InterpreterExecutionScope m_interpreter_scope;
+    JS::ExecutionContext m_execution_context;
+    WeakPtr<JS::Realm> m_worker_realm;
+    RefPtr<WorkerDebugConsoleClient> m_console;
+    JS::GlobalObject* m_worker_scope;
+
+    NonnullRefPtr<MessagePort> m_implicit_port;
+    RefPtr<MessagePort> m_outside_port;
+    RefPtr<MessagePort> m_inside_port;
+
+    void run_a_worker(AK::URL& url, EnvironmentSettingsObject& outside_settings, MessagePort& outside_port, WorkerOptions const& options);
+};
+
+// FIXME: This is a bit ugly, ESO should simply take the document instead of window.
+class WorkerEnvironmentSettingsObject final
+    : public EnvironmentSettingsObject
+    , public Weakable<WorkerEnvironmentSettingsObject> {
+public:
+    WorkerEnvironmentSettingsObject(DOM::Document& document, JS::ExecutionContext& execution_context)
+        : EnvironmentSettingsObject(execution_context)
+        , m_document(document)
+    {
+    }
+    static WeakPtr<WorkerEnvironmentSettingsObject> setup(DOM::Document& document, JS::ExecutionContext& execution_context /* FIXME: null or an environment reservedEnvironment, a URL topLevelCreationURL, and an origin topLevelOrigin */)
```

Comment:
Worker ESO seems to mostly be a copy of the Window ESO. (For example, setup here is talking about window on step 2 and workers don't have a responsible document)
Here's the part of the spec for worker ESO :) https://html.spec.whatwg.org/multipage/workers.html#set-up-a-worker-environment-settings-object

Reply:
That's because it was a stub implementation copied from Window ESO :^) since base ESO is abstract, and I couldn't use WindowESO due to it requiring a Window object, which the inner worker environment won't have.
Should I just remove the WindowESO comments to clean it up and avoid confusion?

Reply:
I went ahead and removed all the WindowESO comments and replaced them with a comment explaining where it comes from and why.